Fluid typography: add font size constraints

ramonjd opened the pull request:
Resolves https://github.com/WordPress/gutenberg/issues/44957

WordPress/wordpress-develop patch:

- https://github.com/WordPress/wordpress-develop/pull/3489

Cherry picked into wp/6.1 branch in this PR:

- https://github.com/WordPress/gutenberg/pull/45133

## What?

This PR adds default minimum font size constraint of 14px/0.875rem/0.875em so that a fluid font's minimum font size does not become smaller than the constraint.

❗ For user-defined font sizes that are < 14px we will use the font-size itself as the minimum value. This is so that changes to these custom values are reflected in the editor and frontend. We trust the user here to insert a sensible value

## Why and how?
To ensure that minimum font sizes, whether user-defined or calculated by `gutenberg_get_typography_font_size_value` are not less than 14px/0.875rem/0.875em.

This is to protect readability and accessibility.

## Testing Instructions
In theme.json, add some very small font sizes (<14px or .875rem) to the presets' `fluid.min`.

Example:

```json
				{
					"size": "2.8rem",
					"fluid": {
						"min": "0.3rem",
						"max": "30px"
					},
					"slug": "neutrale",
					"name": "Neutrale"
				},
```

Add a post and apply the preset.

Check that these font sizes are never less than 14px or .875rem in the editor and frontend.

Next add a custom font size to an element, either in theme.json or via the editor.

If your custom font size is <14px or .875rem, the value of the custom font size will be static.

Otherwise the clamped value's minimum font size should be never be less than 14px or .875rem.

<details>

<summary>Example theme.json</summary>

```json
{
	"version": 2,
	"settings": {
		"typography": {
			"fluid": true,
			"fontSizes": [
				{
					"size": ".9rem",
					"slug": "piccolino",
					"name": "Piccolino"
				},
				{
					"size": "2rem",
					"fluid": {
						"min": "1.8rem",
						"max": "40px"
					},
					"slug": "in-mezzo",
					"name": "In Mezzo"
				},
				{
					"size": "2.8rem",
					"fluid": {
						"min": "0.3rem",
						"max": "30px"
					},
					"slug": "neutrale",
					"name": "Neutrale"
				},
				{
					"size": "4.75rem",
					"slug": "grandone",
					"name": "Grandone"
				}
			]
		}
	},
	"styles": {
		"typography": {
			"fontSize": "12px"
		},
		"elements": {
			"h1": {
				"typography": {
					"fontSize": "4rem"
				}
			}
		},
		"blocks": {
			"core/post-date": {
				"typography": {
					"fontSize": "16px"
				}
			}
		}
	}
}

```

</details>

<details>

<summary>Example HTML</summary>

```html
<!-- wp:heading -->
<h2>Default font size in theme.json style (not clamped because &lt; 14px):</h2>
<!-- /wp:heading -->

<!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet. A eaque harum in aspernatur vero qui sapiente asperiores quo harum exercitationem qui quasi neque et laborum veritatis. Et sint inventore rem aliquam tempora nam ratione quaerat.</p>
<!-- /wp:paragraph -->

<!-- wp:heading -->
<h2>Corrected font size in preset (won't be smaller than 14px):</h2>
<!-- /wp:heading -->

<!-- wp:paragraph {"fontSize":"piccolino"} -->
<p class="has-piccolino-font-size">Lorem ipsum dolor sit amet. A eaque harum in aspernatur vero qui sapiente asperiores quo harum exercitationem qui quasi neque et laborum veritatis. Et sint inventore rem aliquam tempora nam ratione quaerat.</p>
<!-- /wp:paragraph -->

<!-- wp:heading -->
<h2>Corrected fluid.min font size in preset (won't be smaller than 14px):</h2>
<!-- /wp:heading -->

<!-- wp:paragraph {"fontSize":"neutrale"} -->
<p class="has-neutrale-font-size">Lorem ipsum dolor sit amet. A eaque harum in aspernatur vero qui sapiente asperiores quo harum exercitationem qui quasi neque et laborum veritatis. Et sint inventore rem aliquam tempora nam ratione quaerat.</p>
<!-- /wp:paragraph -->

<!-- wp:heading -->
<h2>Small custom font size (not clamped because &lt; 14px):</h2>
<!-- /wp:heading -->

<!-- wp:paragraph {"style":{"typography":{"fontSize":"12px"}}} -->
<p style="font-size:12px">Lorem ipsum dolor sit amet. A eaque harum in aspernatur vero qui sapiente asperiores quo harum exercitationem qui quasi neque et laborum veritatis. Et sint inventore rem aliquam tempora nam ratione quaerat.</p>
<!-- /wp:paragraph -->
```

</details>
## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/6458278/196073775-1459f362-cbc1-4e36-81aa-e46660730ad1.mp4

ramonjd committed:
Initial commit:
- Adds default minimum font size limits so that text does not become smaller than 14px/0.875rem/0.875em. The consequence will be that all fluid font sizes will be no smaller than 14px/0.875rem/0.875em in any viewport width.
- Users may define font-sizes that are smaller than 14px/0.875rem/0.875em. In this case the user-defined font size becomes the minimum. For example if a user defines 9px for a block, then the minimum font size for that block will be 9px.

- Min font size should not be greater than max font size.
- Round to 3 decimal places (JS)

- Rounding computed min and max values

ramonjd committed:
Rounding parsed min and max values

ramonjd committed:
Updating JS rounding function so that values are rounded

ramonjd committed:
Removing type coercion

ramonjd committed:
Updating failing test

ramonjd committed:
Remove reference to preset in favour of "fontSize"

tellthemachines commented on lib/block-supports/typography.php:
```diff
@@ -342,11 +342,11 @@ function gutenberg_get_typography_value_and_unit( $raw_value, $options = array()
+ *     @type string $minimum_viewport_width  Minimum viewport size from which type will have fluidity.
+ *     @type string $maximum_font_size       Maximum font size for any clamp() calculation.
+ *     @type string $minimum_font_size       Minimum font size for any clamp() calculation.
+ *     @type int    $scale_factor            A scale factor to determine how fast a font scales within boundaries.
```
Was this meant to change?

ramonjd replied:
Oops, no. Thanks for flagging it.

tellthemachines commented on lib/block-supports/typography.php:
```diff
@@ -357,11 +357,13 @@ function gutenberg_get_computed_fluid_typography_value( $args = array() ) {
 	$scale_factor               = isset( $args['scale_factor'] ) ? $args['scale_factor'] : null;
 
-	// Grab the minimum font size and normalize it in order to use the value for calculations.
+	// Grabs the minimum font size and normalize it in order to use the value for calculations.
```
Should be "Grabs... and normalizes"

tellthemachines commented on lib/block-supports/typography.php:
```diff
@@ -372,12 +374,17 @@ function gutenberg_get_computed_fluid_typography_value( $args = array() ) {
 	);
 
-	// Protect against unsupported units.
+	// Checks for mandatory min and max sizes, and protects against unsupported units.
```
In the (unchanged) lines above, we're coercing the max size to `$font_size_unit` but not the min size, is that deliberate?

ramonjd replied:
It is, and here's the logic (mine at least 😄)

The min and max font size units can theoretically differ, so one can be `px` and the other `rem`. This is fine in 
the resulting clamp value's first and third args, e.g., `clamp(20px, 1.25rem + ((1vw - 7.68px) * 93.75), 50rem)`

But in order to calculate a linear factor between viewport boundaries we have to do the math on similar values, that is, values whose units are the same. So we coerce the max value unit to whatever the min value unit is.

I use the min value because, if there isn't a min value specified, e.g., it's a custom font size, we use incoming custom font size to calculate a minimum value, and therefore use its unit.

The functions' internals are ripe for refactor by the way, but since it's all neatly housed I think we can do it in a follow up.

tellthemachines commented on lib/block-supports/typography.php:
```diff
@@ -451,51 +458,85 @@ function gutenberg_get_typography_font_size_value( $preset, $should_use_fluid_ty
+	);
+
+	/*
+	 * Checks if a user-defined font size is lower than $minimum_font_size_limit. If so than it should become the minimum font size.
```
Typo: "if so, then" (with or without the comma; I guess we could remove "then" altogether if we use the comma)

andrewserong replied:
Nit: tiny typo, also might be worth splitting onto a new line starting from `If so` since this line is pretty long?

```suggestion
	 * Checks if a user-defined font size is lower than $minimum_font_size_limit. If so then it should become the minimum font size.
```

tellthemachines commented on packages/block-editor/src/components/font-sizes/fluid-utils.js:
```diff
@@ -68,15 +70,54 @@ export function getComputedFluidTypographyValue( {
+		);
+
+		/*
+		 * Checks if a user-defined font size is lower than $minimum_font_size_limit. If so than it should become the minimum font size.
```
Same typo as above: "if so, then"

andrewserong replied:
Same typo as above 😄

```suggestion
		 * Checks if a user-defined font size is lower than $minimum_font_size_limit. If so then it should become the minimum font size.
```

ramonjd committed:
Ensure that rem units are used when calculating min and max boundaries in em
This is to keep it consistent with the original formula and ensure the fluid argument in clamp is working off a size relative to :root

Fixing formatting and grammar
Rolling back renaming $preset to $font_size in preference to a follow up.

Do  not clamp value if there is no min size and the font size is less than 14px

tellthemachines reviewed (commented):
Tested these changes on Chrome, Firefox and Safari on macOS. I used custom font sizes both large and small, defined in px, em and rem, and tested with different browser font size settings. It's all working pretty well across all the browsers!

A few comments below.

andrewserong commented on lib/block-supports/typography.php:
```diff
@@ -413,36 +420,36 @@ function gutenberg_get_computed_fluid_typography_value( $args = array() ) {
- * @param array $preset                     {
- *     Required. fontSizes preset value as seen in theme.json.
+ * @param array $font_size                     {
+ *     Required. fontSizes represented in the preset format as seen in theme.json.
```
Tiny nit: now that the param is `$font_size` what if we reorder the wording here?

```suggestion
 *     Required. A font size as represented in the fontSizes preset format as seen in theme.json.
```

ramonjd replied:
Yes! Thanks. 

Just to note the reason behind the changed arg name: now that any font size can be fluidified, "preset" no longer makes sense.

andrewserong commented on lib/block-supports/typography.php:
```diff
@@ -451,51 +458,85 @@ function gutenberg_get_typography_font_size_value( $preset, $should_use_fluid_ty
+			'minimum_font_size'       => $minimum_font_size_raw,
+			'maximum_font_size'       => $maximum_font_size_raw,
+			'scale_factor'            => $default_scale_factor,
+			'minimum_font_size_limit' => $default_minimum_font_size_limit,
```
I looks like the minimum limit has already been applied, does it still need to be passed to this array?

ramonjd replied:
It does not. 

Let's keep this between us 😆

andrewserong reviewed (commented):
This is looking pretty good so far, thanks so much for digging in to land a fix @ramonjd! Just left a few nits, but happy to give this a more detailed test once it's ready for a final review 🙂

andrewserong commented on lib/block-supports/typography.php:
```diff
@@ -342,11 +342,11 @@ function gutenberg_get_typography_value_and_unit( $raw_value, $options = array()
- *     @type string $maximum_font_size      Maximum font size for any clamp() calculation.
- *     @type string $minimum_font_size      Minimum font size for any clamp() calculation.
- *     @type int    $scale_factor           A scale factor to determine how fast a font scales within boundaries.
+ *     @type string $maximum_viewport_width  Maximum size up to which type will have fluidity.
+ *     @type string $minimum_viewport_width  Minimum viewport size from which type will have fluidity.
+ *     @type string $maximum_font_size       Maximum font size for any clamp() calculation.
+ *     @type string $minimum_font_size       Minimum font size for any clamp() calculation.
+ *     @type int    $scale_factor            A scale factor to determine how fast a font scales within boundaries.
```
Looks like a no-longer needed whitespace change?

ramonjd commented:
> The exception is where users or themers define font-sizes that are smaller than 14px. In this case, in order to honour the user's choice, the value is not clamped (no fluidity).

@jasmussen @tellthemachines @andrewserong 

Seeking your counsel here. 

I've made an assumption about font sizes that are < 14px set by users or in theme.json (essentially any font size without a defined min font size).  

In this PR they are not clamped, but will return a static size. 

So, if the font size is `12px` it won't be clamped, and remain `12px`.

`14px` and above will start returning, for example, `clamp(14px, 0.875rem + (1vw - 7.68px) * 0.841, 21px)` as you go bigger.

In a previous commit, I had returned a clamped value but had used font sizes that are < 14px as the minimum font size. That seemed a bit strange and against the spirit of what we're trying to acheive.

What are your thoughts on how to handle these small values set by the user?

My intention was not to create any surprises in the editor or theme.json, by constraining a font size even if the user is setting the value < 14px. It would be strange to set a font size to `9px` and not seeing the text budge on the screen.

Also I thought that there may be some design cases for very small font sizes, e.g., where the text is not meant to be readable. It's a long bow to draw, I know.

ramonjd committed:
Oh Linter! You persistent crank.

andrewserong commented:
> In this PR they are not clamped, but will return a static size.

From my perspective, the idea of keeping font sizes < `14px` as non clamped values sounds like a good approach to me. Taking a step back and thinking of the value of the fluid typography rules, it's primarily to enable setting larger font-sizes that scale down in a pleasing (and magic) way on smaller viewport sizes, or scale up in a pleasing way to very large viewport sizes. So, when someone explicitly uses a really small font size, I doubt there's the same desire to have fluid rules be applied.

I'm not a designer, though, so keen to hear what @jasmussen thinks here, too!

Also, I suppose in follow-ups (perhaps for 6.2), we might be able to have an explicit `fluid` flag, where if users _really_ want the small font-size to have fluid rules, they could be applied?

But when it comes to really small font-sizes, the main use cases I can think of are footer areas where folks are intentionally putting a whole bunch of information with a high level of density (especially terms of service info and the like). And that information is likely the kind that isn't intended to really scale up very much. E.g. here are parts of footers on Microsoft and Apple websites:

<img width="815" alt="image" src="https://user-images.githubusercontent.com/14988353/196078576-ba777354-ae44-4036-af14-edfff1c753c7.png">

<img width="1024" alt="image" src="https://user-images.githubusercontent.com/14988353/196078611-d110c147-681f-4681-b306-a9c49b7a536c.png">

tellthemachines commented:
> In a previous commit, I had returned a clamped value but had used font sizes that are < 14px as the minimum font size. That seemed a bit strange and against the spirit of what we're trying to acheive.

It doesn't seem strange to me. I think it stranger if the fluid behaviour varies depending on font size, because it feels a bit random - why set the limit at 14px? 

We could also argue that if the user sets a custom font size, they'll expect that size to persist across different viewports, but if that behaviour changes with setting the `fluid` flag, then it makes sense for it to change for all font sizes equally.

andrewserong commented:
> but if that behaviour changes with setting the fluid flag, then it makes sense for it to change for all font sizes equally.

Also a very good point!

ramonjd commented:
> We could also argue that if the user sets a custom font size, they'll expect that size to persist across different viewports, but if that behaviour changes with setting the fluid flag, then it makes sense for it to change for all font sizes equally

> Taking a step back and thinking of the value of the fluid typography rules, it's primarily to enable setting larger font-sizes that scale down in a pleasing (and magic) way on smaller viewport sizes, or scale up in a pleasing way to very large viewport sizes. So, when someone explicitly uses a really small font size, I doubt there's the same desire to have fluid rules be applied.

Hmm, excellent observations. 

So the choice is whether to:

- return a clamped value all the time, even if the font size is < 14px
- only return a clamped value if the font size > 14px (and there is no minimum set in a preset for example)

Both give precedence to user choice, which I think we should do. 

I'm not sure whether we place such a constraint on styles elsewhere so we'd be setting a precedent that should be compatible with other potential constraints, e.g., maximum font size. 

With that in mind, I suppose turning on `fluid: true` is a deliberate act on the part of a theme, so it'd be easy to argue that it should be a "universal" switch.

Keen to hear what @jasmussen and co. think as well.

The good news is that the functionality for both exists in the commit history of this PR, so it's easy to change once we've made up our minds 😄 

Later I'd like to revive this PR:

- https://github.com/WordPress/gutenberg/pull/42489

And add a configurable minimum font size as well. Maybe even a max font size.

jasmussen reviewed (approved):
Nice! Approving from a design perspective, because when `fluid: true;` and no other fluid boundaries defined on a per-preset basis, the minimum font size is 14px in editor and frontend:
<img width="265" alt="Screenshot 2022-10-17 at 10 41 20" src="https://user-images.githubusercontent.com/1204802/196131850-66390c8b-933e-4a4f-b157-cbb8cb0fb595.png">

<img width="343" alt="Screenshot 2022-10-17 at 10 41 28" src="https://user-images.githubusercontent.com/1204802/196131862-fa02190a-3ff4-4eed-8866-3905e51da8bd.png">

That definitely fixes the small-font bug.

I furthermore tested that the manual boundaries (custom min/max) work as intended, and even fluid: false on a per-preset basis works as intended. This makes fluid typography feel true to its boolean "just works" intent, but maintains the room to customize.

I'd appreciate good code reviews in addition to this approval! Thank you for fast work.

jasmussen commented:
> Taking a step back and thinking of the value of the fluid typography rules, it's primarily to enable setting larger font-sizes that scale down in a pleasing (and magic) way on smaller viewport sizes, or scale up in a pleasing way to very large viewport sizes. So, when someone explicitly uses a really small font size, I doubt there's the same desire to have fluid rules be applied.

Just to respond to this one, here's my understanding of how this works:

* If I add `fluid: true;` to settings.typography, all my presets are by default fluid, with a min-value of 14px.
* If I add `flud: false;` to the same place, they are not fluid.
* If I add `flud: false;` to a single preset, it is not fluid.
* If I add custom min/max values to a single preset, those values are used.
* The 14px minimum, I expect, acts as a _floor_ to the values, not as a blanket replacement for all clamp min-values. That is, if you add a `"size": "0.75rem"` to your theme.json, a value which would extrapolate to clamp values between (just as a made up example) say 8px and 18px, then then in effect it would clamp between 14px and 18px. If yes, this is fine, because presumably you could set fluid: false, or provide custom min/max values.

Is that right?

ramonjd commented:
Thanks for the quick review and your thoughts @jasmussen 

The following answers describe the "is" state, including that of this PR. Not the "should be" state 😄 

> If I add fluid: true; to settings.typography, all my presets are by default fluid, with a min-value of 14px.

Yes and no. 

The way this PR works right now is that wherever a font size is defined, and only a font size, which means no specified fluid min/max values, the font size will only be clamped if it is =>14px.

Otherwise it will be static. 

So, even in a preset, if `"size": "0.4rem",` is defined only, then the size will be `0.4rem`.

The reason for this behaviour is that I thought it struck a good balance between:

1. how much control we give the user and 
2. protecting the generated clamp functions from having minimum font sizes that are too small.

~Now that I've had a sleep, I think we could make an exception for presets and always have a floor of 14px regardless of the circumstances. The reason being is that one can add `false` to a font size preset and turn fluidity off for any preset.~ I'm no longer sure about this.

What do folks think?

> If I add flud: false; to the same place, they are not fluid.

Yes!

> If I add flud: false; to a single preset, it is not fluid.

Exactly.

> If I add custom min/max values to a single preset, those values are used.

Yes. And if the min value is <14px we bump it up to 14px.

> The 14px minimum, I expect, acts as a floor to the values, not as a blanket replacement for all clamp min-values. That is, if you add a "size": "0.75rem" to your theme.json, a value which would extrapolate to clamp values between (just as a made up example) say 8px and 18px, then then in effect it would clamp between 14px and 18px. If yes, this is fine, because presumably you could set fluid: false, or provide custom min/max values.

Yes. 

What I'm trying to preserve however is the following: when a user applies a single font size to an element, e.g.,  a font size that is less than 14px, whether via styles in the theme.json or in the editor and, they should see that font size represented in the editor and frontend.

Right now a clamped value will never have a minimum value of <14px, but a user can still set a smaller font size if they want. It just won't be clamped.

I would consider it a bug if I really really wanted a 5px sized-font (fluid or not) and the software forced me up to 14px without any recourse or explanation.

Given the speed at which this change _might_ make it to Core, I think we should tread carefully here.

The other option is that we can clamp font sizes <14px using the user-defined <14px value as the minimum.

Curious to hear folks' opinions. Thank you!

andrewserong commented:
Thanks for laying that out @ramonjd, and for the feedback @jasmussen!

> Given the speed at which this change might make it to Core, I think we should tread carefully here.
> The other option is that we can clamp font sizes <14px using the user-defined <14px value as the minimum.

Yes, I think whichever way we go, we need to preserve the ability for someone to manually set font-sizes on individual blocks to < `14px` when `fluid: true` is set overall for the theme, otherwise the custom control itself will feel broken.

I don't feel too strongly about whether that < `14px` results in fluid being switched off for that particular font size, or if it sets the floor of the fluid calculation. But I think @tellthemachines made a good argument for preserving fluid logic as long as we have the one global boolean that opts-in to the feature.

In the spirit of treading carefully, perhaps the most consistent approach is to go with a font size of < `14px` setting the floor of the `clamp()` rules? That then opens up a neat path for 6.2 where we might have a boolean control in the UI to switch off `clamp()` in the Typography panel if desired.

ramonjd committed:
For font sizes of < 14px that have no defined minimum sizes, use the font size to set the floor of the clamp() value.

ramonjd committed:
Removing max < min check. It wasn't there before. It can be a follow up if required as it requires extra logic.

ramonjd commented:
> In the spirit of treading carefully, perhaps the most consistent approach is to go with a font size of < 14px setting the floor of the clamp() rules? That then opens up a neat path for 6.2 where we might have a boolean control in the UI to switch off clamp() in the Typography panel if desired.

Thanks @andrewserong. I've reverted back to clampifying all values as you and @tellthemachines advise. 

So, for those catching up, the present state in this PR is, when fluid typography is set to `true` at the global level:

### 💠 We will attempt to convert all single font sizes to clamp() values.

For user-defined font sizes that are < 14px we will use the font-size itself as the minimum value. This is so that changes to these custom values are reflected in the editor and frontend. We trust the user here to insert a sensible value.

Example:

```json
{
	"size": "2px",
	"slug": "piccolino",
	"name": "Piccolino"
}
```

Becomes:

`--wp--preset--font-size--piccolino: clamp(2px, 0.125rem + ((1vw - 7.68px) * 0.12), 3px);`

The same is true for values from the editor/theme.json styles. So `"2px"` becomes `clamp(2px, 0.125rem + ((1vw - 7.68px) * 0.12), 3px)`.

### 💠 Wherever a supplied min value (that is, a min value from a preset's `fluid.min` or computed internally by the function) is < 14px the clamp() value's min argument will be corrected upwards.

```json
{
	"size": "2.8rem",
	"fluid": {
		"min": "0.5rem",
		"max": ".75rem"
	},
	"slug": "neutrale",
	"name": "Neutrale"
},
```

Becomes:

`--wp--preset--font-size--neutrale: clamp(0.875rem, 0.875rem + ((1vw - 0.48rem) * -0.24), .75rem)`

-----

```json
{
	"size": "0.875em",
	"slug": "in-mezzo",
	"name": "In Mezzo"
},
```

Becomes:

`--wp--preset--font-size--in-mezzo: clamp(0.875em, 0.875rem + ((1vw - 0.48em) * 0.842), 1.313em);`

Because normally we'd compute the min value using `min * DEFAULT_MINIMUM_FONT_SIZE_FACTOR (0.75)`

### 💠 (Already in 6.1) Individual font size presets with `"fluid": false` will not be clampified.


```json
{
	"size": "4.75rem",
	"fluid": false,
	"slug": "grandone",
	"name": "Grandone"
}
```

Becomes:

`--wp--preset--font-size--grandone: 4.75rem;`

-----

### Other options

1. We force a floor of 14px everywhere, no exceptions
2. We force a floor of 14px for presets only. Custom font sizes will not be constrained.
3. We return the static font size for all font sizes/min font sizes < 14px, and only convert to clamp where a font size/min font size > 14px.

andrewserong commented on packages/block-editor/src/components/font-sizes/fluid-utils.js:
```diff
@@ -66,46 +76,95 @@ export function getComputedFluidTypographyValue( {
+		if ( !! minimumFontSizeLimitParsed?.value ) {
+			/*
+			 * If a minimum size was not passed to this function
+			 * and the user-defined font size is lower than $minimum_font_size_limit,
```
Tiny nit: remove reference to PHP variable. This is totally optional as the meaning is still clear, I'd only worry about this if you have other changes to commit, too

```suggestion
			 * and the user-defined font size is lower than the minimum font size limit,
```

andrewserong commented on packages/block-editor/src/components/font-sizes/fluid-utils.js:
```diff
@@ -66,46 +76,95 @@ export function getComputedFluidTypographyValue( {
+
+				/*
+				 * Otherwise, if the passed or calculated minimum font size is lower than $minimum_font_size_limit
+				 * use $minimum_font_size_limit instead.
```
As above, tiny nit: remove references to PHP variables. Totally optional, don't worry about making this change if you don't have others to make, too

andrewserong commented on packages/block-editor/src/components/font-sizes/fluid-utils.js:
```diff
@@ -133,17 +192,20 @@ export function getComputedFluidTypographyValue( {
 		( ( maximumFontSizeParsed.value - minimumFontSizeParsed.value ) /
 			( maximumViewPortWidthParsed.value -
 				minumumViewPortWidthParsed.value ) );
-	linearFactor = roundToPrecision( linearFactor, 3 ) || 1;
```
This used to have a fallback of `1`, I imagine to avoid multiplying by `0`. Can the above calculation still theoretically result in `0`? E.g. if the max font size value is the same as the minimum font size value, or if the max viewport width is the same as the minimum viewport width?

E.g, if I set `min` and `max` to the same value in a preset (not very useful, I know), then the output will result in a `* 0`:

<img width="418" alt="image" src="https://user-images.githubusercontent.com/14988353/196350076-df6f0689-2b7a-4af8-bbd3-bb6f713a6229.png">

In that case, prior to this PR the multiplier would have been `* 1`.

I think we can resolve this here by updating the below lines to:

```js
	const linearFactorScaled = roundToPrecision(
		( linearFactor || 1 ) * scaleFactor,
		3
	);
```

ramonjd replied:
Oh thanks. That nearly slipped through. 🤭 

I'll add a test as well

ramonjd replied:
Or, we could bail early when the min and max are the same. Having a clamped value in that case, as you hint at, is not very useful 😄

andrewserong commented on packages/block-editor/src/components/font-sizes/fluid-utils.js:
```diff
@@ -215,7 +289,8 @@ export function getTypographyValueAndUnit( rawValue, options = {} ) {
+	const base = Math.pow( 10, digits + 1 );
 	return Number.isFinite( value )
-		? parseFloat( value.toFixed( digits ) )
+		? parseFloat( ( value + 1 / base ).toFixed( digits ) )
```
It looks like this adds a `1` at a single decimal place smaller than the rounding digits. I take it this is to ensure we work around weird floating point precision issues? I see a slightly different approach has been taken elsewhere in the repo earlier (e.g. in [the performance reporter](https://github.com/WordPress/gutenberg/blob/45cffa2beaedd9146944f3a084b3748ce5572362/packages/e2e-tests/config/performance-reporter.js#L12)), but this one looks good to me.

ramonjd replied:
Yes, it was to ensure we get 3 all the time, even when JS randomly throws `0.000001` at us :D 

Thanks for linking to the other one. It should work as well. Might be a good opportunity later to extract the util

andrewserong reviewed (approved):
This is testing nicely for me @ramonjd, thanks for all the follow-up.

In terms of a "works out of the box" approach, the changes involved here feel most consistent for me, without us offering up additional flags for controlling fluid typography when a user is selecting a custom font size (via a checkbox or other means). Playing around with the fluid typography, in the editor, I think this strikes a good balance, though personally I look forward to being able to adjust the multiplier / ratio further down the track (e.g. in 6.2) to take a slightly more moderate calculation — but that's a feature for another time!

✅ The floor value of `14px` appears to be in use everywhere that's expected (in presets, in `rem` and `em` values, in custom font sizes set at the individual level and in global styles)
✅ When a font-size value of < that `14px` is used, then the provided value is set correctly as the floor in all the above cases I tested it with. This to me feels like a good justification for landing this fix, as without it, setting small custom font-sizes results in very small clampified values — a good common use case for small font-sizes is terms & conditions like information that appears an a lot of sites (particularly those with signups or selling products)

I just left a comment on one small issue in the JS calculations that would be good to fix up before merging — if the `min` and `max` values are the same, then the multiplier winds up being `* 0`. Looks like hopefully a quick fix, and also probably not a _huge_ issue as in practice the min and max values will always be different, or there's not much use in using the fluid feature!

Other than that, this LGTM! ✨

ramonjd committed:
Checking for a zero-based linear factor. If we find one, default to `1`

ramonjd committed:
Refer to correct JS var in JS file

ramonjd commented:
Thanks again for testing @andrewserong 

I've made all the changes and added test for the zero value.

> Playing around with the fluid typography, in the editor, I think this strikes a good balance, though personally I look forward to being able to adjust the multiplier / ratio further down the track (e.g. in 6.2) to take a slightly more moderate calculation — but that's a feature for another time!

I'm also happy with the balance so far and I think we're in a good place. Existing font sizes, when a user flips the fluid switch won't be affected and we can introduce configurables later.

I'm grateful for the constant testing and thoughts! 🙇

ockham commented:
I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: c3933faacf

ockham commented:
> I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: [c3933fa](https://github.com/WordPress/gutenberg/commit/c3933faacf4aa654dcad8c3d7d271e36616851db)

Reverting on the `wp/6.1` branch per [this discussion](https://wordpress.slack.com/archives/C02RQBWTW/p1666711047763149).

ockham commented:
> > I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: [c3933fa](https://github.com/WordPress/gutenberg/commit/c3933faacf4aa654dcad8c3d7d271e36616851db)
> 
> Reverting on the `wp/6.1` branch per [this discussion](https://wordpress.slack.com/archives/C02RQBWTW/p1666711047763149).

9b745d7f57
